hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java

```diff
@@ -0,0 +1,214 @@
+    this.conf = conf;
+    executorService = HadoopExecutors.newScheduledThreadPool(
+        this.conf.getInt(OzoneConfigKeys.OZONE_SCM_CONTAINER_THREADS,
+            OzoneConfigKeys.OZONE_SCM_CONTAINER_THREADS_DEFAULT),
+        new ThreadFactoryBuilder().setDaemon(true)
+            .setNameFormat("Container State Machine Thread - %d").build());
+    connectionManager = new SCMConnectionManager(
+        OzoneClientUtils.getScmRpcTimeOutInMilliseconds(conf),
+        conf);
+    context = new StateContext(this.conf, DatanodeStates.getInitState(), this);
+    taskWaitTime = this.conf.getLong(OzoneConfigKeys.OZONE_CONTAINER_TASK_WAIT,
+        OzoneConfigKeys.OZONE_CONTAINER_TASK_WAIT_DEFAULT);
+    heartbeatFrequency = OzoneClientUtils.getScmHeartbeatInterval(conf);
+  }
+
+  /**
+   * Returns the Connection manager for this state machine.
+   *
+   * @return - SCMConnectionManager.
+   */
+  public SCMConnectionManager getConnectionManager() {
+    return connectionManager;
+  }
+
+  /**
+   * Runs the state machine at a fixed frequency.
+   */
+  public void start() throws IOException {
+    long now = 0;
+    long nextHB = 0;
+    while (context.getState() != DatanodeStates.SHUTDOWN) {
+      try {
+        nextHB = Time.monotonicNow() + heartbeatFrequency;
+        context.execute(executorService, taskWaitTime, TimeUnit.SECONDS);
+        now = Time.monotonicNow();
+        if (now < nextHB) {
+          Thread.sleep(nextHB - now);
+        }
+      } catch (InterruptedException | ExecutionException | TimeoutException e) {
+        LOG.error("Unable to finish the execution", e);
+      }
+    }
+  }
+
+  /**
+   * Gets the current context.
+   *
+   * @return StateContext
+   */
+  public StateContext getContext() {
+    return context;
+  }
+
+  /**
+   * Sets the current context.
+   *
+   * @param context - Context
+   */
+  public void setContext(StateContext context) {
+    this.context = context;
+  }
+
+  /**
+   * Closes this stream and releases any system resources associated with it. If
+   * the stream is already closed then invoking this method has no effect.
+   * <p>
+   * <p> As noted in {@link AutoCloseable#close()}, cases where the close may
+   * fail require careful attention. It is strongly advised to relinquish the
+   * underlying resources and to internally <em>mark</em> the {@code Closeable}
+   * as closed, prior to throwing the {@code IOException}.
+   *
+   * @throws IOException if an I/O error occurs
+   */
+  @Override
+  public void close() throws IOException {
+    executorService.shutdown();
+    try {
+      if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
+        executorService.shutdownNow();
+      }
+
+      if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
+        LOG.error("Unable to shutdown statemachine properly.");
+      }
+    } catch (InterruptedException e) {
+      executorService.shutdownNow();
+      Thread.currentThread().interrupt();
```

Comment:
Can we add a Log.error() after the Thread.currentThread.interrupt() for troubleshooting of shutdown issues in the future?